### ISSUE TITLE
Simplification of API regarding collections

### ISSUE THREAD

**Issue by brianreavis** [naturalatlas/node-gdal#14]
I tried to do an audit on the main GDAL/OGR features that represent arrays / collections. This is partially related to #10. Here are some thoughts on what we can and can't do with them.
### Standardized Language
- "delete", "remove", "unset" → "remove"
- "add", "push", "create" → "add"
- "length", "count" → "count"
### Consideration: Potentially-Async Features

While GDAL is not particularly thread-safe at the moment, we shouldn't corner ourselves into assuming certain features will always have to be synchronous.

> [GDAL Wiki](https://trac.osgeo.org/gdal/wiki/FAQMiscellaneous#IstheGDALlibrarythread-safe): **Is the GDAL library thread-safe?**
> 
> No. GDAL is not completely thread safe.
> 
> However for GDAL 1.3.0 much work has been done on making some common scenarios thread safe. In particular for the situation where many threads are reading from GDAL datasets at once should work as long as no two threads access the same  GDALDataset object at the same time. However, in this scenario, no threads can be writing to GDAL while others are reading or chaos may ensue.
> 
> Also, while the GDAL core infrastructure is now thread-safe for this specific case, only a few drivers have been vetted to be thread safe. It is intended that work will continue on improving GDAL's thread safety in future versions.

For the features that are potentially async, we can't get too clever like we could if we assumed it was sync:

``` js
// syntactically-nice array-like objects won't work:
rasterband.overviews[index] // getter for getOverview
rasterband.overviews.length // getter for getOverviewCount
```
#### RasterBand Collections (potentially-async)

``` js
gdal.Dataset.prototype.getRasterBand(i)
gdal.Dataset.prototype.getRasterCount()
```

This could be normalized to:

``` js
dataset.bands.count()
dataset.bands.get(index)
dataset.bands.forEach(callback)
```
#### Layer Collections (potentially-async)

``` js
ogr.Datasource.prototype.getLayerCount()
ogr.Datasource.prototype.getLayer(i)
ogr.Datasource.prototype.deleteLayer(i)
ogr.Datasource.prototype.createLayer(string)
ogr.Datasource.prototype.getLayerByName(i)
```

This could be normalized to:

``` js
datasource.layers.count()
datasource.layers.get(name|index)
datasource.layers.remove(name|index)
datasource.layers.add(layer)
datasource.layers.forEach(callback)
```
#### Feature Collections (potentially-async)

``` js
ogr.Layer.prototype.getFeatureCount()
ogr.Layer.prototype.getFeature(i)
ogr.Layer.prototype.getNextFeature()
ogr.Layer.prototype.setFeature(i)
ogr.Layer.prototype.createFeature()
ogr.Layer.prototype.deleteFeature(i)
```

This could be normalized to:

``` js
layer.features.count()
layer.features.get(index)
layer.features.remove(index)
layer.features.add(feature)
layer.features.forEach(callback)
```
#### Driver Collections

``` js
gdal/ogr.getDriver()
gdal/ogr.getDriverCount
```

This could be simplified to:

``` js
[gdal|ogr].drivers.count()
[gdal|ogr].drivers.get(name|index)
```
#### Field Collections

``` js
ogr.Feature.prototype.getFieldCount()
ogr.Feature.prototype.getFieldIndex()
ogr.Feature.prototype.getFieldAs[X]()
ogr.Feature.prototype.isFieldSet(i)
ogr.Feature.prototype.unsetField(i)
ogr.Feature.prototype.getField(i)
ogr.Feature.prototype.setField(i)
ogr.Feature.prototype.getFields()
ogr.Feature.prototype.setFields()
```

This could be simplified to:

``` js
feature.fields.count()
feature.fields.reset({})
feature.fields.set({})
feature.fields.set(key|index, value)
feature.fields.isset(key|index)
feature.fields.remove(key|index)
feature.fields.get(key|index, [type])
feature.fields.toJSON()
```

This removes the need for all of these:

``` js
- ogr.Feature.prototype.getFieldAsInteger()
- ogr.Feature.prototype.getFieldAsDouble()
- ogr.Feature.prototype.getFieldAsString()
- ogr.Feature.prototype.getFieldAsIntegerList()
- ogr.Feature.prototype.getFieldAsDoubleList()
- ogr.Feature.prototype.getFieldAsStringList()
- ogr.Feature.prototype.getFieldAsBinary()
- ogr.Feature.prototype.getFieldAsDateTime()
```
#### Field Definition Collections

``` js
ogr.FeatureDefn.prototype.getFieldCount()
ogr.FeatureDefn.prototype.getFieldDefn(i)
ogr.FeatureDefn.prototype.addFieldDefn()
ogr.FeatureDefn.prototype.deleteFieldDefn(i)
ogr.FeatureDefn.prototype.reorderFieldDefns()
```

This could be simplified to:

``` js
layer.def.fields.count()
layer.def.fields.add(field_defn)
layer.def.fields.remove(index)
layer.def.fields.get(index)

// "mapping" could be a map of keys to indexes, or indexes to indexes
layer.def.fields.reorder(mapping)
```
### TL;DR:
- Decide on consistent language to be used across the binding ("add", "remove", "count") 
- Have better encapsulation of functionality. Example: GDAL groups layer CRUD methods in with all the other DataSource methods that don't really have to do with layers. Given a data source is a representation of a collection of layers + some other things, I think it makes sense to have `layers` be its own thing in `DataSource`. I think it's cleaner & more intuitive.
